gobblin-runtime/src/main/java/gobblin/runtime/spec_store/FSSpecStore.java

```diff
@@ -0,0 +1,273 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gobblin.runtime.spec_store;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URI;
+import java.nio.charset.Charset;
+import java.util.Collection;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.typesafe.config.Config;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.runtime.api.GobblinInstanceEnvironment;
+import gobblin.runtime.api.Spec;
+import gobblin.runtime.api.SpecNotFoundException;
+import gobblin.runtime.api.SpecStore;
+import gobblin.util.PathUtils;
+
+
+/**
+ * The Spec Store for file system to persist the Spec information.
+ * Note:
+ * 1. This implementation has no support for caching.
+ * 2. This implementation does not performs implicit version management.
+ *    For implicit version management, please use a wrapper FSSpecStore.
+ */
+@Slf4j
+public class FSSpecStore implements SpecStore {
+
+  private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+  protected final Config sysConfig;
+  protected final FileSystem fs;
+  protected final String fsSpecStoreDir;
+  protected final Path fsSpecStoreDirPath;
+
+  public FSSpecStore(GobblinInstanceEnvironment env)
+      throws IOException {
+    this(env.getSysConfig().getConfig());
+  }
+
+  public FSSpecStore(Config sysConfig)
+      throws IOException {
+    Preconditions.checkArgument(sysConfig.hasPath(ConfigurationKeys.SPECSTORE_FS_DIR_KEY),
+        "FS SpecStore path must be specified.");
+
+    this.sysConfig = sysConfig;
+    this.fsSpecStoreDir = this.sysConfig.getString(ConfigurationKeys.SPECSTORE_FS_DIR_KEY);
+    this.fsSpecStoreDirPath = new Path(this.fsSpecStoreDir);
+    try {
+      this.fs = this.fsSpecStoreDirPath.getFileSystem(new Configuration());
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to detect job config directory file system: " + e, e);
+    }
+  }
+
+  @Override
+  public boolean exists(URI specUri) throws IOException {
```

Comment:
It seems like in cases like delete the check should be for the full file path, so we should be able to use the underlying fs.exists() instead of listing and matching on the prefix.